src/node/timer.h

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "ds/logger.h"
+#include "ds/spinlock.h"
+
+#include <chrono>
+#include <functional>
+#include <list>
+
+namespace ccf
+{
+  using TimerCallback = std::function<bool()>;
+
+  /**
+   * A timer class to trigger actions periodically.
+   *
+   * A shared_ptr to a new Timer is returned by Timers::new_timer(period, cb).
+   * The timer is only active when it is started (start()) and marked as expired
+   * when the callback is triggered. If the callback returns true, the timer
+   * continues ticking. Otherwise, the timer expires and will tick again only
+   * when it is explicitely re-started (start()).
+   *
+   **/
+  class Timer
+  {
+  private:
+    enum TimerState
+    {
+      STOPPED = 0,
+      STARTED,
+      EXPIRED
+    };
+
+    SpinLock lock;
+    std::chrono::milliseconds period;
+    std::chrono::milliseconds tick_;
```

Comment:
Why `tick_`? We use this in a few places to distinguish argument `foo_` from member `foo`, but this is a member?

Reply:
Ah, because we have a method with the same name. I think this member should be called `elapsed`

Reply:
Much better indeed!